oxd-server/src/main/java/org/gluu/oxd/server/op/Validator.java

```diff
@@ -164,7 +158,38 @@ public void validateIdToken(String nonce, String clientId) {
         }
     }
 
+    //Added this method so that we can write the test cases
+    public static void validateAudience(Jwt idToken, String clientId) {
+
+        final String audienceFromToken = idToken.getClaims().getClaimAsString(JwtClaimName.AUDIENCE);
+        if (!clientId.equalsIgnoreCase(audienceFromToken)) {
+            List<String> audAsList = idToken.getClaims().getClaimAsStringList(JwtClaimName.AUDIENCE);
+            if (audAsList != null) {
+                if (!audAsList.stream().anyMatch(aud -> clientId.equalsIgnoreCase(aud))) {
+                    LOG.error("ID Token has invalid audience (string list). Expected audience: " + clientId + ", audience from token is: " + audAsList);
+                    throw new HttpException(ErrorResponseCode.INVALID_ID_TOKEN_BAD_AUDIENCE);
+                }
+                //If the ID Token contains multiple audiences, the Client SHOULD verify that an azp Claim is present.
+                if (audAsList.size() > 1) {
+                    String azpFromToken = idToken.getClaims().getClaimAsString(JwtClaimName.AUTHORIZED_PARTY);
+                    //If an azp (authorized party) Claim is present, the Client SHOULD verify that its client_id is the Claim Value. If present, it MUST contain the OAuth 2.0 Client ID of this party.
+                    if (!Strings.isNullOrEmpty(azpFromToken) && !azpFromToken.equalsIgnoreCase(clientId)) {
+                        LOG.error("ID Token has invalid authorized party (string list). Expected authorized party: " + clientId + ", authorized party from token is: " + azpFromToken);
+                        throw new HttpException(ErrorResponseCode.INVALID_ID_TOKEN_BAD_AUTHORIZED_PARTY);
+                    }
+                }
+            }
+
+            // somehow fetching string list does not return actual list, so we do this ugly trick to compare single valued array, more details in #178
```

Comment:
Can you check https://github.com/GluuFederation/oxd/issues/178 still works? After commenting this code I believe it doesn't.

Reply:
Made code changes and added test cases to validate.